File: backend/src/dtos.rs
use std::collections::HashMap;

use chrono::{DateTime, Local};
use serde::{Deserialize, Serialize};

use crate::models::{Board, BoardId, Column, ColumnId, Task, TaskId};

#[derive(Debug, Serialize)]
pub struct BoardDTO {
    pub board: Board,
    pub columns: HashMap<ColumnId, Column>,
    pub tasks: HashMap<TaskId, Task>,
}

#[derive(Debug, Deserialize)]
pub struct TaskCreationPayload {
    pub id: TaskId,
    pub name: String,
    pub description: String,
    pub due_date: Option<DateTime<Local>>,
    pub position: f64,
    pub column_id: ColumnId,
}

#[derive(Debug, Deserialize)]
pub struct TaskDumpingPayload {
    pub to_column: ColumnId,
}

#[derive(Debug, Deserialize)]
pub struct TaskPatchPayload {
    pub name: Option<String>,
    pub description: Option<String>,
    #[serde(default)]
    pub update_due_date: bool,
    pub due_date: Option<DateTime<Local>>,
    pub column_id: Option<String>,
    pub position: Option<f64>,
}

#[derive(Debug, Deserialize)]
pub struct ColumnPatchPayload {
    pub name: Option<String>,
    pub position: Option<f64>,
    pub description: Option<String>,
    pub completes_tasks: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct ColumnCreationPayload {
    pub id: ColumnId,
    pub name: String,
    pub description: String,
    pub position: f64,
    pub board_id: BoardId,
    pub completes_tasks: bool,
}

#[derive(Debug, Deserialize)]
pub struct ColumnDumpingPayload {
    pub to: ColumnId,
}

#[derive(Debug, Deserialize)]
pub struct BoardCreationPayload {
    pub id: BoardId,
    pub name: String,
    pub description: String,
}

#[derive(Debug, Deserialize)]
pub struct BoardPatchPayload {
    pub name: Option<String>,
    pub description: Option<String>,
    pub pinned: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct RoutinePatchPayload {
    pub name: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ScheduledActivityPatchPayload {
    pub name: Option<String>,
    #[serde(default)]
    pub update_task_id: bool,
    pub task_id: Option<TaskId>,
    pub starts_on: Option<DateTime<Local>>,
    pub duration_minutes: Option<i64>,
}

File: frontend/src/types/dtos.ts
import type { Board, Column, Task, ScheduledActivity, Routine, RoutineActivity } from "./models"

export type Bootstrap = {
  boards: Record<string, Board>
  columns: Record<string, Column>
  tasks: Record<string, Task>
  scheduled_activities: Record<string, ScheduledActivity>
  routines: Record<string, Routine>
  routine_activities: Record<string, RoutineActivity>
}

export type BoardDto = {
  board: Board
  columns: Record<string, Column>
  tasks: Record<string, Task>
}

export type TaskCreationPayload = {
  id: string
  name: string
  description: string
  due_date?: string
  position: number
  column_id: string
}

export type TaskDumpPayload = {
  to_column: string
  to_position: number
}

export type TaskPatchPayload = {
  title?: string
  description?: string
  column_id?: string,
  update_due_date: boolean,
  due_date?: string,
  position?: number
}

export type ColumnPatchPayload = {
  name?: string
  description?: string,
  position?: number
  completes_tasks?: boolean
}

export type ColumnDumpingPayload = {
  to: string
}

export type TaskDumpingPayload = {
  to_column: string
}

export type ColumnCreationPayload = {
  id: string
  name: string
  position: number
  description: string,
  board_id: string
  completes_tasks: boolean
}

export type BoardCreationPayload = {
  id: string
  name: string
  description: string,
}

export type BoardPatchPayload = {
  name?: string
  description?: string,
  pinned?: boolean
}

export type RoutinePatchPayload = {
  name?: string
}

export type ScheduledActivityPatchPayload = {
  name: string
  task_id?: string | null
  starts_on?: Date
  duration_minutes: number
}
